pkg/drivers/krunkit/krunkit.go

```diff
@@ -0,0 +1,561 @@
+	return fmt.Errorf("hosts without a driver cannot be upgraded")
+}
+
+func (d *Driver) getKrunkitState() (state.State, error) {
+	pidfile := d.pidfilePath()
+	pid, err := process.ReadPidfile(pidfile)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return state.Error, err
+		}
+		return state.Stopped, nil
+	}
+	exists, err := process.Exists(pid, driverName)
+	if err != nil {
+		return state.Error, err
+	}
+	if !exists {
+		// No process, stale pidfile.
+		if err := os.Remove(pidfile); err != nil {
+			log.Debugf("failed to remove %q: %s", pidfile, err)
+		}
+		return state.Stopped, nil
+	}
+	return state.Running, nil
+}
+
+func (d *Driver) stopKrunkit() error {
+	if err := d.SetKrunkititState("Stop"); err != nil {
+		// krunkit may be already stopped, shutting down, or not listening.
+		log.Debugf("Failed to set krunkit state to 'Stop': %s", err)
+		pidfile := d.pidfilePath()
+		pid, err := process.ReadPidfile(pidfile)
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return err
+			}
+			// No pidfile.
+			return nil
+		}
+		if err := process.Terminate(pid, driverName); err != nil {
+			if err != os.ErrProcessDone {
+				return err
+			}
+			// No process, stale pidfile.
+			if err := os.Remove(pidfile); err != nil {
+				log.Debugf("failed to remove %q: %s", pidfile, err)
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+func (d *Driver) killkrunkit() error {
+	if err := d.SetKrunkititState("HardStop"); err != nil {
+		// Typically fails with EOF due to https://github.com/crc-org/krunkit/issues/277.
+		log.Debugf("Failed to set krunkit state to 'HardStop': %s", err)
+		pidfile := d.pidfilePath()
+		pid, err := process.ReadPidfile(pidfile)
+		if err != nil {
+			if !errors.Is(err, os.ErrNotExist) {
+				return err
+			}
+			// No pidfile.
+			return nil
+		}
+		if err := process.Kill(pid, driverName); err != nil {
+			if err != os.ErrProcessDone {
+				return err
+			}
+			// No process, stale pidfile.
+			if err := os.Remove(pidfile); err != nil {
+				log.Debugf("failed to remove %q: %s", pidfile, err)
+			}
+			return nil
+		}
+	}
+	return nil
+}
+
+func (d *Driver) publicSSHKeyPath() string {
+	return d.GetSSHKeyPath() + ".pub"
+}
+
+func (d *Driver) diskPath() string {
+	return d.ResolveStorePath("disk.img")
+}
+
+func (d *Driver) pidfilePath() string {
+	return d.ResolveStorePath(pidFileName)
+}
+
+func (d *Driver) logfilePath() string {
+	return d.ResolveStorePath(logFileName)
+}
+
+func (d *Driver) serialPath() string {
+	return d.ResolveStorePath(serialFileName)
+}
+
+func (d *Driver) isoPath() string {
+	return d.ResolveStorePath(isoFileName)
+}
+
+func (d *Driver) restfulURI() string {
+	// TODO: use unused port or a unix socket when supported.
+	// https://github.com/containers/krunkit/issues/47
+	return "tcp://localhost:8081"
+}
+
+func (d *Driver) vmStateURI() string {
+	return d.restfulURI() + "/vm/state"
+}
+
+// Make a boot2docker VM disk image.
+func (d *Driver) generateDiskImage(size int) error {
```

Comment:
Duplicate of vfkit version, which may duplicate of qemu?

Reply:
if this is an exact duplicate, lets add a TODO: issue to make them in a reusable code shared across 3 drirvers, we already have code in the Parent Driver package that is shared across all drivers

Reply:
The main reason was that qemu (and thus vfkit) were external drivers to libmachine, which didn't have helpers.

For qemu2 and minikube-machine, there are more opportunities to add it to the common driver SDK package

Reply:
Added comment linking to #21090